libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -321,6 +318,81 @@ string YulUtilFunctions::shiftRightFunctionDynamic()
 	});
 }
 
+string YulUtilFunctions::shiftRightSignedFunctionDynamic()
+{
+	string const functionName = "shift_right_signed_dynamic";
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(bits, value) -> result {
+				<?hasShifts>
+					result := sar(bits, value)
+				<!hasShifts>
+					let divisor := exp(2, bits)
+					switch slt(value, 0)
+					case 0 {
+						result := div(value, divisor)
+					}
+					default {
+						result := not(div(not(value), divisor))
+					}
+				</hasShifts>
+			}
+			)")
+			("functionName", functionName)
+			("hasShifts", m_evmVersion.hasBitwiseShifting())
+			.render();
+	});
+}
+
+
+string YulUtilFunctions::typedShiftLeftFunction(Type const& _type, bool _amountSigned)
+{
+	solAssert(_type.category() == Type::Category::FixedBytes || _type.category() == Type::Category::Integer, "");
+	string const functionName = "shift_left_" + _type.identifier() + (_amountSigned ? "_signedAmount" : "");
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(value, bits) -> result {
```

Comment:
A bit confusing that the order is reversed compared to the others and evm shifts here, isn't it?

Reply:
Yes - the idea was that this is the higher-level function that is closer to the operator...

Reply:
Hm... ok, that's an argument I guess.